File: RELINKBACKEND.py
from flask import Flask, send_from_directory, request
from flask_socketio import SocketIO, join_room, leave_room, emit

app = Flask(__name__, static_folder='public')
socketio = SocketIO(app, cors_allowed_origins="*")  # allow all origins

rooms = {}

@app.route('/')
def index():
    return send_from_directory('public', 'index.html')

@socketio.on('setUsername')
def handle_set_username(name):
    emit('usernameSet')

@socketio.on('createRoom')
def handle_create_room(room_name):
    join_room(room_name)
    if room_name not in rooms:
        rooms[room_name] = []
    rooms[room_name].append(request.sid)
    emit('roomJoined', room_name)
    emit('userList', rooms[room_name], room=room_name)

@socketio.on('sendMessage')
def handle_message(msg):
    for room_name, users in rooms.items():
        if request.sid in users:
            emit('newMessage', {'username': 'User', 'msg': msg}, room=room_name)

@socketio.on('disconnect')
def handle_disconnect():
    for room_name, users in rooms.items():
        if request.sid in users:
            users.remove(request.sid)
            emit('userList', users, room=room_name)

if __name__ == '__main__':
    socketio.run(app, host='0.0.0.0', port=int(os.environ.get('PORT', 5000)))

File: RELINKBACKEND.js
// Minimal RELINK chat prototype (Render-ready)
// Backend: Node.js + Express + Socket.IO

const express = require('express');
const http = require('http');
const { Server } = require('socket.io');

const app = express();
const server = http.createServer(app);
const io = new Server(server);

app.use(express.static('public')); // Serve frontend files

// Store rooms and users in memory
const rooms = {};

io.on('connection', (socket) => {
  let currentRoom = null;
  let username = null;

  socket.on('setUsername', (name) => {
    username = name;
    socket.emit('usernameSet', name);
  });

  socket.on('createRoom', (roomName) => {
    if (!rooms[roomName]) {
      rooms[roomName] = [];
    }
    socket.join(roomName);
    currentRoom = roomName;
    rooms[roomName].push(username);
    io.to(currentRoom).emit('userList', rooms[currentRoom]);
    socket.emit('roomJoined', roomName);
  });

  socket.on('joinRoom', (roomName) => {
    if (rooms[roomName]) {
      socket.join(roomName);
      currentRoom = roomName;
      rooms[roomName].push(username);
      io.to(currentRoom).emit('userList', rooms[currentRoom]);
      socket.emit('roomJoined', roomName);
    } else {
      socket.emit('error', 'Room does not exist');
    }
  });

  socket.on('sendMessage', (msg) => {
    if (currentRoom) {
      io.to(currentRoom).emit('newMessage', { username, msg });
    }
  });

  socket.on('disconnect', () => {
    if (currentRoom && rooms[currentRoom]) {
      rooms[currentRoom] = rooms[currentRoom].filter(user => user !== username);
      io.to(currentRoom).emit('userList', rooms[currentRoom]);
    }
  });
});

// Use Render's PORT environment variable if it exists
const PORT = process.env.PORT || 3000;
server.listen(PORT, () => {
  console.log(`RELINK backend running on port ${PORT}`);
});
